### PR TITLE
Document Torque class types and %-style intrinsics

### PR DESCRIPTION
**Pull request by dannoc**

### CODE REVIEW

File: src/docs/torque.md [hunk 1]
```diff
@@ -197,6 +198,58 @@ type int32 generates 'TNode<Int32T>' constexpr 'int32_t';
...
+
+ClassFieldDeclaration: <b>weak</b><sub>opt</sub> FieldDeclaration;
+
+FieldDeclaration: Identifier <b>:</b> Type <b>;</b>
+</code></pre>
+
+An example class:
+
+```torque
+class JSFunction extends JSObject' {
```
Comment (mathiasbynens) [inline src/docs/torque.md]: typo: `'`

File: src/docs/torque.md [hunk 2]
```diff
@@ -197,6 +198,58 @@ type int32 generates 'TNode<Int32T>' constexpr 'int32_t';
...
+  feedback_cell: Smi;
+  weak code: Code;
+  weak prototype_or_initial_map: Object | Map;
+}
+```
+
+On the Torque side, the field declarations in classes implicitly generate field getters and setters, e.g.:
+
+```torque
+operator ‘.shared_function_info’ macro LoadJSFunctionSharedFunctionInfo(JSFunction): SharedFunctionInfo;
```
Comment (mathiasbynens) [inline src/docs/torque.md]: I’m guessing these single quotes should be straight instead of curly?

File: src/docs/torque.md [hunk 3]
```diff
@@ -289,12 +343,55 @@ There is only one copy of the code for a Torque builtin, and that is in the gene
...
+// %FromConstexpr converts a constexpr value into into a non-constexpr
+// value. Currently, only conversion to the following non-constexpr types
+// are supported: Smi, Number, String, uintptr, intptr, and int32
+intrinsic %FromConstexpr<To: type, From: type>(b: From): To;
+
+// %Allocate allocates an unitialized object of size 'size' from V8's
+// GC heap and "reinterpret casts" the resuting object pointer to the
+// specified Torque class, allowing constructors to subsequently use
+// standard field access operators to initialize the object.
+// This intrinsic should never be called from Torque code. It used
```
Comment (mathiasbynens) [inline src/docs/torque.md]: Typo: `It used` → `It’s used`

File: src/docs/torque.md [hunk 4]
```diff
@@ -289,12 +343,55 @@ There is only one copy of the code for a Torque builtin, and that is in the gene
...
+// %RawConstexprCast translate to static_casts in the generated C++ code.
+intrinsic %RawConstexprCast<To: type, From: type>(f: From): To;
+
+// %FromConstexpr converts a constexpr value into into a non-constexpr
+// value. Currently, only conversion to the following non-constexpr types
+// are supported: Smi, Number, String, uintptr, intptr, and int32
+intrinsic %FromConstexpr<To: type, From: type>(b: From): To;
+
+// %Allocate allocates an unitialized object of size 'size' from V8's
+// GC heap and "reinterpret casts" the resuting object pointer to the
```
Comment (mathiasbynens) [inline src/docs/torque.md]: Typo: resulting

File: src/docs/torque.md [hunk 5]
```diff
@@ -289,12 +343,55 @@ There is only one copy of the code for a Torque builtin, and that is in the gene
...
+// specified Torque class, allowing constructors to subsequently use
+// standard field access operators to initialize the object.
+// This intrinsic should never be called from Torque code. It used
+// internally when desugaring the 'new' operator.
+intrinsic %Allocate<Class: type>(size: intptr): Class;
+```
+
+Like `buildtin`s and `runtime`s, `intrinsic`s cannot have labels.
```
Comment (mathiasbynens) [inline src/docs/torque.md]: typo: `builtin`

File: src/docs/torque.md [hunk 6]
```diff
@@ -231,12 +284,13 @@ There are two special types indicated by the keywords `void` and `never`. `void`
 
 ### Callables
 
-Callables are conceptually like functions in JavaScript or C++, but they have some additional semantics that allow them to interact in useful ways with CSA code and with the V8 runtime. Torque provides three types of callables: `macro`s, `builtin`s, and `runtime`s.
+Callables are conceptually like functions in JavaScript or C++, but they have some additional semantics that allow them to interact in useful ways with CSA code and with the V8 runtime. Torque provides three types of callables: `macro`s, `builtin`s, `runtime`s and `intrinsics`s.
```
[the remark this answers is not in the document]
Comment (jakobkummerow) [reply src/docs/torque.md]: also: s/three/four/ :-)

[the remark this answers is not in the document]
Comment (mathiasbynens) [reply src/docs/torque.md]: off-by-one!

**Pull request merged**